crates/core/tedge_api/src/cmd_topic.rs

```diff
@@ -0,0 +1,151 @@
+use mqtt_channel::Topic;
+use mqtt_channel::TopicFilter;
+
+#[derive(Debug, Clone, Eq, PartialEq)]
+pub enum DeviceKind {
+    Main,
+    Child(String),
+}
+
+impl DeviceKind {
+    /// Return the device name, "main" if DeviceKind::Main, otherwise child ID
+    pub fn name(&self) -> String {
+        self.name_with_default("main")
+    }
+
+    /// Return the device name, the given input if DeviceKind::Main, otherwise child ID
+    pub fn name_with_default(&self, name: &str) -> String {
+        match self {
+            DeviceKind::Main => name.to_string(),
+            DeviceKind::Child(child_id) => child_id.to_string(),
+        }
+    }
+}
+
+#[derive(Debug, Clone, Eq, PartialEq)]
+pub struct Target {
+    device_id: String,
+    cmd_id: String,
+}
+
+impl Target {
+    pub fn new(device_id: String, cmd_id: String) -> Self {
+        Target { device_id, cmd_id }
+    }
+}
+
+#[derive(Debug, Clone, Eq, PartialEq)]
+pub enum CmdPublishTopic {
+    // Restart(Target),
+    // SoftwareList(Target),
+    // SoftwareUpdate(Target),
+    // ConfigSnapshot(Target),
+    // ConfigUpdate(Target),
+    LogUpload(Target),
+}
+
+impl CmdPublishTopic {
+    pub fn to_topic(&self, prefix: &str) -> Topic {
+        let topic = match self {
+            CmdPublishTopic::LogUpload(target) => {
+                format!(
+                    "{}/device/{}///cmd/log_upload/{}",
+                    prefix, target.device_id, target.cmd_id
+                )
+            }
+        };
+        Topic::new_unchecked(&topic)
+    }
+}
+
+#[derive(Debug, Clone, Eq, PartialEq)]
+pub enum CmdSubscribeTopic {
+    // Restart,
+    // SoftwareList,
+    // SoftwareUpdate,
+    // ConfigSnapshot,
+    // ConfigUpdate,
+    LogUpload,
+}
+
+impl CmdSubscribeTopic {
+    pub fn metadata(&self, prefix: &str) -> String {
+        match self {
+            CmdSubscribeTopic::LogUpload => format!("{prefix}/device/+///cmd/log_upload"),
+        }
+    }
+
+    pub fn with_id(&self, prefix: &str) -> String {
+        format!("{}/+", self.metadata(prefix))
+    }
+}
+
+/// Return device ID and command ID
+pub fn get_target_ids_from_cmd_topic(
+    topic: &Topic,
+    prefix: &str,
+) -> (Option<DeviceKind>, Option<String>) {
+    let mut cmd_topic_filter =
+        TopicFilter::new_unchecked(&format!("{prefix}/device/+/+/+/cmd/+/+"));
+    cmd_topic_filter.add_unchecked(&format!("{prefix}/device/+/+/+/cmd/+"));
```

Comment:
We should not assume a scheme for the entity identifiers.

```suggestion
        TopicFilter::new_unchecked(&format!("{prefix}/+/+/+/+/cmd/+/+"));
    cmd_topic_filter.add_unchecked(&format!("{prefix}/+/+/+/+/cmd/+"));
```